moveit_core/robot_state/test/robot_state_benchmark.cpp

```diff
@@ -66,312 +61,279 @@ Eigen::Isometry3d createTestIsometry()
-static void moveItJacobian(benchmark::State& st)
+BENCHMARK_DEFINE_F(RobotStateBenchmark, jacobianMoveIt)(benchmark::State& st)
 {
-  // Load a test robot model.
-  const moveit::core::RobotModelPtr& robot_model = moveit::core::loadTestingRobotModel(PANDA_TEST_ROBOT);
-
-  // Make sure the group exists, otherwise exit early with an error.
-  if (!robot_model->hasJointModelGroup(PANDA_TEST_GROUP))
+  moveit::core::RobotState state(robot_model);
+  const moveit::core::JointModelGroup* jmg = state.getJointModelGroup(PANDA_TEST_GROUP);
+  if (!jmg)
   {
     st.SkipWithError("The planning group doesn't exist.");
     return;
   }
 
-  // Robot state.
-  moveit::core::RobotState kinematic_state(robot_model);
-  const moveit::core::JointModelGroup* jmg = kinematic_state.getJointModelGroup(PANDA_TEST_GROUP);
-
-  // Provide our own random number generator to setToRandomPositions to get a deterministic sequence of joint
-  // configurations.
+  // Manually seeded RandomNumberGenerator for deterministic results
   random_numbers::RandomNumberGenerator rng(0);
 
   for (auto _ : st)
   {
     // Time only the jacobian computation, not the forward kinematics.
     st.PauseTiming();
-    kinematic_state.setToRandomPositions(jmg, rng);
-    kinematic_state.updateLinkTransforms();
+    state.setToRandomPositions(jmg, rng);
+    state.updateLinkTransforms();
     st.ResumeTiming();
-    kinematic_state.getJacobian(jmg);
+    state.getJacobian(jmg);
   }
 }
 
 // Benchmark time to compute the Jacobian using KDL.
-static void kdlJacobian(benchmark::State& st)
+BENCHMARK_DEFINE_F(RobotStateBenchmark, jacobianKDL)(benchmark::State& st)
 {
-  const moveit::core::RobotModelPtr& robot_model = moveit::core::loadTestingRobotModel(PANDA_TEST_ROBOT);
-
-  // Make sure the group exists, otherwise exit early with an error.
-  if (!robot_model->hasJointModelGroup(PANDA_TEST_GROUP))
+  moveit::core::RobotState state(robot_model);
+  const moveit::core::JointModelGroup* jmg = state.getJointModelGroup(PANDA_TEST_GROUP);
+  if (!jmg)
   {
     st.SkipWithError("The planning group doesn't exist.");
     return;
   }
-
-  // Robot state.
-  moveit::core::RobotState kinematic_state(robot_model);
-  const moveit::core::JointModelGroup* jmg = kinematic_state.getJointModelGroup(PANDA_TEST_GROUP);
-
-  // Provide our own random number generator to setToRandomPositions to get a deterministic sequence of joint
-  // configurations.
-  random_numbers::RandomNumberGenerator rng(0);
-
   KDL::Tree kdl_tree;
   if (!kdl_parser::treeFromUrdfModel(*robot_model->getURDF(), kdl_tree))
   {
     st.SkipWithError("Can't create KDL tree.");
     return;
   }
 
-  KDL::Chain kdl_chain;
-  if (!kdl_tree.getChain(jmg->getJointModels().front()->getParentLinkModel()->getName(),
-                         jmg->getLinkModelNames().back(), kdl_chain))
-  {
-    st.SkipWithError("Can't create KDL Chain.");
-    return;
-  }
+  KDL::TreeJntToJacSolver jacobian_solver(kdl_tree);
```

Comment:
Curious on the difference between `TreeJntToJacSolver` and `ChainJntToJacSolver`?
Is it a different implementation, or actually calling the same implementation underneath?

Reply:
These are two different classes implementing the same code essentially. The central difference is the size of the Jacobian: tree considers columns for all joints in the tree, chain only those in the chain.